File: daily/1122.数组的相对排序.js
/*
 * @lc app=leetcode.cn id=1122 lang=javascript
 *
 * [1122] 数组的相对排序
 */

// @lc code=start
/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number[]}
 */
var relativeSortArray = function(arr1, arr2) {
  const map = new Map()
  const right = []
  for (let n of arr2) map.set(n, 0)
  for (let n of arr1)
    if (map.has(n)) map.set(n, map.get(n) + 1)
    else right.push(n)
  const arr = []
  for (let n of arr2) {
    let count = map.get(n)
    while(count--) arr.push(n)
  }
  arr.push(...right.sort((a, b) => a - b))
  return arr
}
// @lc code=end

File: daily/1122.数组的相对排序.rs
/*
 * @lc app=leetcode.cn id=1122 lang=rust
 *
 * [1122] 数组的相对排序
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn relative_sort_array(arr1: Vec<i32>, arr2: Vec<i32>) -> Vec<i32> {
        let mut map = HashMap::new();
        let mut right = Vec::new();
        let mut result = Vec::new();
        for i in 0..arr2.len() {
            map.insert(arr2[i], 0);
        }
        for i in 0..arr1.len() {
            if map.contains_key(&arr1[i]) {
                let count = map.entry(arr1[i]).or_insert(0);
                *count += 1;
            } else {
                right.push(arr1[i]);
            }
        }
        for i in 0..arr2.len() {
            let count = map.get(&arr2[i]).unwrap();
            for j in 0..*count {
                result.push(arr2[i]);
            }
        }
        right.sort();
        result.append(&mut right);
        return result;
    }
}
// @lc code=end

